Sources/_StringProcessing/Engine/Processor.swift

```diff
@@ -730,12 +700,23 @@ extension String {
     _ scalar: Unicode.Scalar,
     at pos: Index,
     limitedBy end: String.Index,
-    boundaryCheck: Bool
+    boundaryCheck: Bool,
+    isCaseInsensitive: Bool
   ) -> Index? {
+    // TODO: extremely quick-check-able
+    // TODO: can be sped up with string internals
     assert(end <= endIndex)
 
-    guard pos < end, unicodeScalars[pos] == scalar else {
-      return nil
+    guard pos < end else { return nil }
+    let curScalar = unicodeScalars[pos]
+
+    if isCaseInsensitive {
+      guard curScalar.properties.lowercaseMapping == scalar.properties.lowercaseMapping
+      else {
+        return nil
+      }
+    } else {
+      guard curScalar == scalar else { return nil }
     }
 
     let idx = unicodeScalars.index(after: pos)
```

Comment:
I can't add a comment to the next line, but I think it should be an assertion instead of a test… e.g. `assert(idx <= end)`. If `pos < end` (as checked above), then `unicodeScalars.index(after: pos)` can't be greater than `end` unless something is malformed.

Reply:
It can if `end` is a sub-scalar index, such as pointing to the second byte of a multi-byte scalar